src/garage/sampler/rl2_sampler.py

```diff
@@ -0,0 +1,195 @@
+"""BatchSampler which uses VecEnvExecutor to run multiple environments."""
+from collections import OrderedDict
+import itertools
+import pickle
+import time
+
+from dowel import logger, tabular
+import numpy as np
+
+from garage.experiment import deterministic
+from garage.misc import tensor_utils
+from garage.misc.prog_bar_counter import ProgBarCounter
+from garage.sampler.batch_sampler import BatchSampler
+from garage.sampler.stateful_pool import singleton_pool
+from garage.sampler.utils import truncate_paths
+from garage.sampler.vec_env_executor import VecEnvExecutor
+
+
+class RL2Sampler(BatchSampler):
+    """BatchSampler which uses VecEnvExecutor to run multiple environments.
+
+    This sampler is for RL^2. See https://arxiv.org/pdf/1611.02779.pdf.
+
+    Args:
+        algo (garage.np.algos.RLAlgorithm): An algorithm instance.
+        env (garage.envs.GarageEnv): An environement instance.
+        meta_batch_size (int): Number of meta batch for sampling.
+        episode_per_task (int): Number of episode for each task.
+        n_envs (int): Number of environment instances to setup.
+            This parameter has effect on sampling performance.
+
+    """
+
+    def __init__(self,
+                 algo,
+                 env,
+                 meta_batch_size,
+                 episode_per_task,
+                 n_envs=None):
+        if n_envs is None:
+            n_envs = singleton_pool.n_parallel * 4
+        super().__init__(algo, env)
+        self._n_envs = n_envs
+
+        self._meta_batch_size = meta_batch_size
+        self._episode_per_task = episode_per_task
+        self._vec_env = None
+        self._env_spec = self.env.spec
+
+    def start_worker(self):
+        """Start workers."""
+        n_envs = self._n_envs
+        assert n_envs >= self._meta_batch_size, (
+            'Number of vectorized environments '
+            'should be at least meta_batch_size.')
+        assert n_envs % self._meta_batch_size == 0, (
+            'Number of vectorized '
+            'environments should be a multiple of meta_batch_size.')
+        envs_per_worker = n_envs // self._meta_batch_size
+
+        tasks = self.env.sample_tasks(self._meta_batch_size)
+        vec_envs = []
+        for task in tasks:
+            vec_env = pickle.loads(pickle.dumps(self.env))
+            vec_env.set_task(task)
+            vec_envs.extend([vec_env] * envs_per_worker)
```

Comment:
`[vec_env] * envs_per_worker` This will make a list of references to vec_env, not actually copies. Is it intended?

Reply:
Good catch, it's indeed a careless bug. I wasn't aware of it since I have been using `envs_per_work=1`. Thanks for pointing out.